### ISSUE TITLE
Seeing no benefit using local models with first quickstart example

### ISSUE THREAD

**Issue by satyaloka93** [zou-group/textgrad#68]
I am testing using local models, openai endpoint is llama-server (version: 3303 (51d2ebad)), with the following gguf models:

Meta-Llama-3-8B-Instruct - q8_0
Mistral-7B-Instruct-v0.2 - q8_0
gemma2-27b-it - q4_k_m

Once the models are loaded with llama-server, I access the local endpoint:

```python
client = OpenAI(base_url="http://localhost:8080/v1", api_key="sk-xxxxxxx")
engine = ChatExternalClient(client=client, model_string='mistral7b')
tg.set_backward_engine(engine, override=True)
model = tg.BlackboxLLM(engine)
```
The first quickstart example is set up:

```python
question_string = ("If it takes 1 hour to dry 25 shirts under the sun, "
                   "how long will it take to dry 30 shirts under the sun? "
                   "Reason step by step")

question = tg.Variable(question_string,
                       role_description="question to the LLM",
                       requires_grad=False)

answer = model(question)
answer
```
As expected, before optimization it fails:
* So, it will take approximately 1.2 hours to dry 30 shirts under the sun.

Then the optimization/evaluation code:

```python
answer.set_role_description("concise and accurate answer to the question")

# Step 2: Define the loss function and the optimizer, just like in PyTorch!
# Here, we don't have SGD, but we have TGD (Textual Gradient Descent)
# that works with "textual gradients".
optimizer = tg.TGD(parameters=[answer])
evaluation_instruction = (f"Here's a question: {question_string}. "
                           "Evaluate any given answer to this question, "
                           "be smart, logical, and very critical. "
                           "Just provide concise feedback.")


# TextLoss is a natural-language specified loss function that describes
# how we want to evaluate the reasoning.
loss_fn = tg.TextLoss(evaluation_instruction)

# Step 3: Do the loss computation, backward pass, and update the punchline.
# Exact same syntax as PyTorch!
loss = loss_fn(answer)
loss.backward()
optimizer.step()
answer
```
And I get:
* Therefore, the answer is 1.2 hours.

The above was with Lllama 3, and with the other two I get either 1.2, or even 2 hours as the final answer. Is there something that would preclude all these models from having no benefit from this textgrad example?

**Comment by mertyg** [comment]
It suggests that the loss function + model combinations do not seem to do enough for this particular question. This is a tricky question for these models (that is why even GPT-4o gets wrong in the first place), seems to be beyond the existing capabilities; the local models you tried may require further supervision to get this right.

Just like in any machine learning setting, I'd try to play around with the loss function. e.g., maybe you can make the self-evaluation instruction a bit more targeted, something like "look for errors"; or something. Certainly a lot to do here to explore the design space of loss functions / algorithms!

**Comment by mertyg** [comment]
Interesting, thanks for the note!
Closing this for now, feel free to reopen it if you have any other issues.

**Comment by HelloWorldLTY** [comment]
Hi, thanks for your great question. May I know more about your openai chat setting? I tried to copy your codes and run it, but failed because of connection error. Do we need extra setting? Thanks.

**Comment by satyaloka93** [comment]
> Hi, thanks for your great question. May I know more about your openai chat setting? I tried to copy your codes and run it, but failed because of connection error. Do we need extra setting? Thanks.

Hey, you can look into any openai compliant server, there are many (vLLM, localai, etc.). For llama.cpp, I am running the llama-server on a GGUF quant model (Mistral 7b), reference: https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md.
